From 9d17e9ed6affa2fe5690cadb884e134cba452561 Mon Sep 17 00:00:00 2001
From: Andy Yan <andy.yan@rock-chips.com>
Date: Wed, 18 Nov 2020 10:55:25 +0800
Subject: [PATCH] drm/rockchip: vop2: Disable afbc auto gating

Some resolution(eg: 4096x2049) can't display when AFBC auto gating
enabled.

Change-Id: Iadd2db1e600c7c0bae10ead0851476f2cf3bbe34
Signed-off-by: Andy Yan <andy.yan@rock-chips.com>
---
 drivers/gpu/drm/rockchip/rockchip_drm_vop.h  | 1 +
 drivers/gpu/drm/rockchip/rockchip_drm_vop2.c | 1 +
 drivers/gpu/drm/rockchip/rockchip_vop2_reg.c | 1 +
 drivers/gpu/drm/rockchip/rockchip_vop_reg.h  | 4 ++++
 4 files changed, 7 insertions(+)

diff --git a/drivers/gpu/drm/rockchip/rockchip_drm_vop.h b/drivers/gpu/drm/rockchip/rockchip_drm_vop.h
index d8f26ec0ed83..ea7a178276f0 100644
--- a/drivers/gpu/drm/rockchip/rockchip_drm_vop.h
+++ b/drivers/gpu/drm/rockchip/rockchip_drm_vop.h
@@ -311,6 +311,7 @@ struct vop_afbc {
 	struct vop_reg enable;
 	struct vop_reg win_sel;
 	struct vop_reg format;
+	struct vop_reg auto_gating_en;
 	struct vop_reg rotate;
 	struct vop_reg block_split_en;
 	struct vop_reg pic_vir_width;
diff --git a/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c b/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c
index fad0697ec305..c01b60576dd7 100644
--- a/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c
+++ b/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c
@@ -1628,6 +1628,7 @@ static void vop2_plane_atomic_update(struct drm_plane *plane, struct drm_plane_s
 		afbc_half_block_en = vop2_afbc_half_block_enable(vpstate);
 		VOP_AFBC_SET(vop2, win, enable, 1);
 		VOP_AFBC_SET(vop2, win, format, afbc_format);
+		VOP_AFBC_SET(vop2, win, auto_gating_en, 0);
 		VOP_AFBC_SET(vop2, win, block_split_en, 0);
 		VOP_AFBC_SET(vop2, win, half_block_en, afbc_half_block_en);
 		VOP_AFBC_SET(vop2, win, hdr_ptr, vpstate->yrgb_mst);
diff --git a/drivers/gpu/drm/rockchip/rockchip_vop2_reg.c b/drivers/gpu/drm/rockchip/rockchip_vop2_reg.c
index 755ddbc643b1..fb3249c062cf 100644
--- a/drivers/gpu/drm/rockchip/rockchip_vop2_reg.c
+++ b/drivers/gpu/drm/rockchip/rockchip_vop2_reg.c
@@ -553,6 +553,7 @@ static const struct vop2_layer_data rk3568_vop_layers[] = {
 static const struct vop_afbc rk3568_vop_afbc = {
 	.enable = VOP_REG(RK3568_CLUSTER0_CTRL, 0x1, 1),
 	.format = VOP_REG(RK3568_CLUSTER0_WIN0_AFBCD_CTRL, 0x1f, 2),
+	.auto_gating_en = VOP_REG(RK3568_CLUSTER0_WIN0_AFBCD_OUTPUT_CTRL, 0x1, 4),
 	.half_block_en = VOP_REG(RK3568_CLUSTER0_WIN0_AFBCD_CTRL, 0x1, 7),
 	.block_split_en = VOP_REG(RK3568_CLUSTER0_WIN0_AFBCD_CTRL, 0x1, 8),
 	.hdr_ptr = VOP_REG(RK3568_CLUSTER0_WIN0_AFBCD_HDR_PTR, 0xffffffff, 0),
diff --git a/drivers/gpu/drm/rockchip/rockchip_vop_reg.h b/drivers/gpu/drm/rockchip/rockchip_vop_reg.h
index cb7c3dc5e179..78da19fdb9bb 100644
--- a/drivers/gpu/drm/rockchip/rockchip_vop_reg.h
+++ b/drivers/gpu/drm/rockchip/rockchip_vop_reg.h
@@ -1158,6 +1158,7 @@
 #define RK3568_CLUSTER0_WIN0_DSP_INFO		0x1024
 #define RK3568_CLUSTER0_WIN0_DSP_ST		0x1028
 #define RK3568_CLUSTER0_WIN0_SCL_FACTOR_YRGB	0x1030
+#define RK3568_CLUSTER0_WIN0_AFBCD_OUTPUT_CTRL	0x1050
 #define RK3568_CLUSTER0_WIN0_AFBCD_ROTATE_MODE	0x1054
 #define RK3568_CLUSTER0_WIN0_AFBCD_HDR_PTR	0x1058
 #define RK3568_CLUSTER0_WIN0_AFBCD_VIR_WIDTH	0x105C
@@ -1175,6 +1176,7 @@
 #define RK3568_CLUSTER0_WIN1_DSP_INFO		0x10A4
 #define RK3568_CLUSTER0_WIN1_DSP_ST		0x10A8
 #define RK3568_CLUSTER0_WIN1_SCL_FACTOR_YRGB	0x10B0
+#define RK3568_CLUSTER0_WIN1_AFBCD_OUTPUT_CTRL	0x10D0
 #define RK3568_CLUSTER0_WIN1_AFBCD_ROTATE_MODE	0x10D4
 #define RK3568_CLUSTER0_WIN1_AFBCD_HDR_PTR	0x10D8
 #define RK3568_CLUSTER0_WIN1_AFBCD_VIR_WIDTH	0x10DC
@@ -1194,6 +1196,7 @@
 #define RK3568_CLUSTER1_WIN0_DSP_INFO		0x1224
 #define RK3568_CLUSTER1_WIN0_DSP_ST		0x1228
 #define RK3568_CLUSTER1_WIN0_SCL_FACTOR_YRGB	0x1230
+#define RK3568_CLUSTER1_WIN0_AFBCD_OUTPUT_CTRL	0x1250
 #define RK3568_CLUSTER1_WIN0_AFBCD_ROTATE_MODE	0x1254
 #define RK3568_CLUSTER1_WIN0_AFBCD_HDR_PTR	0x1258
 #define RK3568_CLUSTER1_WIN0_AFBCD_VIR_WIDTH	0x125C
@@ -1211,6 +1214,7 @@
 #define RK3568_CLUSTER1_WIN1_DSP_INFO		0x12A4
 #define RK3568_CLUSTER1_WIN1_DSP_ST		0x12A8
 #define RK3568_CLUSTER1_WIN1_SCL_FACTOR_YRGB	0x12B0
+#define RK3568_CLUSTER1_WIN1_AFBCD_OUTPUT_CTRL	0x12D0
 #define RK3568_CLUSTER1_WIN1_AFBCD_ROTATE_MODE	0x12D4
 #define RK3568_CLUSTER1_WIN1_AFBCD_HDR_PTR	0x12D8
 #define RK3568_CLUSTER1_WIN1_AFBCD_VIR_WIDTH	0x12DC
-- 
2.35.3

